### PR TITLE
GG-29610 - fixed the problem when UnknownHostException in TcpDiscover…

### PR DESCRIPTION
**Pull request by aealeksandrov**
…yVmIpFinder fails node start if one of the addresses can't be resolved to IP

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java [hunk 1]
```diff
@@ -225,7 +225,18 @@ else if (ipStr.indexOf(':') >= 0)
 
                     Collection<InetSocketAddress> res = new ArrayList<>();
 
-                    InetAddress[] inetAddresses = InetAddress.getAllByName(addrStr);
+                    InetAddress[] inetAddresses;
+
+                    try{
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:230]: Whitespace is desired

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java [hunk 2]
```diff
@@ -235,15 +246,28 @@ else if (ipStr.indexOf(':') >= 0)
...
+                catch (IllegalArgumentException e) {
                     throw new IgniteSpiException(errMsg, e);
                 }
             }
             else {
                 try {
                     int port = Integer.parseInt(portStr);
 
-                    return Collections.singleton(new InetSocketAddress(addrStr, port));
+                    Collection<InetSocketAddress> col = new LinkedHashSet<>();
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:257]: I wonder if we really need this if-else here. Can we have same logic block for range and single port?

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java [hunk 3]
```diff
@@ -235,15 +246,28 @@ else if (ipStr.indexOf(':') >= 0)
...
+
+                        for (InetAddress addrs : inetAddresses)
+                            col.add(new InetSocketAddress(addrs, port));
+                    }
+                    catch (UnknownHostException ignored) {
+                        //save previous behavior on UnknownHostException
+                        col = Collections.singleton(new InetSocketAddress(addrStr, port));
+                    }
+
+                    return col;
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:270]: What about multicast finder, btw? It has the same addresses input.

**Review by alamar — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java (hunk 3 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:270]: The current fix is about VmIpFinder. Probably we should fix multicast too but ut should be done in a separate ticket.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java (hunk 1 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java (hunk 2 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java]: The code was updated. Please check one more time.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java [hunk 4]
```diff
@@ -214,40 +214,43 @@ else if (ipStr.indexOf(':') >= 0)
...
-                    int port1 = Integer.parseInt(portStr.substring(0, portStr.indexOf("..")));
-                    int port2 = Integer.parseInt(portStr.substring(portStr.indexOf("..") + 2, portStr.length()));
+                port1 = Integer.parseInt(portStr.substring(0, portStr.indexOf("..")));
+                port2 = Integer.parseInt(portStr.substring(portStr.indexOf("..") + 2, portStr.length()));
+            } else
+                port1 = port2 = Integer.parseInt(portStr);
 
-                    if (port2 < port1 || port1 == port2 || port1 <= 0 || port2 <= 0)
-                        throw new IgniteSpiException(errMsg);
+            if (port1 != port2 && port2 < port1 || port1 <= 0 || port2 <= 0)
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:225]: Can you put the first two into ()? if ((port1 != port2 && port2 < port1) || port1 <= 0 || port2 <= 0) for better readability

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java [hunk 5]
```diff
@@ -214,40 +214,43 @@ else if (ipStr.indexOf(':') >= 0)
...
             String portStr = tokens[1];
+            int port1, port2;
 
             if (portStr.contains("..")) {
-                try {
-                    int port1 = Integer.parseInt(portStr.substring(0, portStr.indexOf("..")));
-                    int port2 = Integer.parseInt(portStr.substring(portStr.indexOf("..") + 2, portStr.length()));
+                port1 = Integer.parseInt(portStr.substring(0, portStr.indexOf("..")));
+                port2 = Integer.parseInt(portStr.substring(portStr.indexOf("..") + 2, portStr.length()));
+            } else
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java:222]: Need line feed before else

**Review by alamar — CHANGES_REQUESTED** [review]
Otherwise, LGTM

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java (hunk 4 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java (hunk 5 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ipfinder/vm/TcpDiscoveryVmIpFinder.java]: Fixed

**Pull request closed without merging**